### ISSUE TITLE
AbstractPathEffect should implement deepcopy or frozen

### ISSUE THREAD

**Issue by gdowding** [matplotlib/matplotlib#3056]
I hit this while working on https://github.com/yhat/ggplot/issues/288

`AbstractPathEffect` is composed of subclass of `TransformNode` which raises an exception during `deepcopy()` with the suggestion to use `frozen()`, but there is no `frozen()` in `AbstractPathEffect`.

```
from matplotlib.patheffects import AbstractPathEffect
from copy import deepcopy
ape = AbstractPathEffect()
deepcopy(ape)

...
    121         raise NotImplementedError(
    122             "TransformNode instances can not be copied. " +
--> 123             "Consider using frozen() instead.")
    124     __deepcopy__ = __copy__
    125 
```

but,

```
ape.frozen()
AttributeError: 'AbstractPathEffect' object has no attribute 'frozen'
```

I'm not familiar with `TransformNode`, so this may be a naive question, but why not implement `__deepcopy__` as `frozen()` in that class?

**Comment by ncsaba** [comment]
I've got the same error when using the theme_xkcd() in ggplot (which under the hood uses matplotlib). There is an attempt in theme_xkcd to catch that exception and just skip copying the TransformNodes - but I still get the exception from another part of the code, so I ended up doing:

# monkey patching TransformNode:
def self_copy(self, *args, **kwargs):
    return self
TransformNode.__deepcopy__ = self_copy

For all those who seek for a quick workaround, that should work (it did work for me).

**Comment by derekbrokeit** [comment]
I had the same problem attempting to run `copy.deepcopy` on a `Figure` object in python3. There should be a easy way to make copies of these objects. Right now it seems that the easiest way to make a deep copy of these objects is using `pickle.loads(pickle.dumps(...))`. This is a scary, not obvious way of writing these operations, so there should be a way to do this. If I am missing something, I would greatly appreciate some guidance.

**Comment by tacaswell** [comment]
Why are you copying figure objects?  I do not think that this will be the only hold of for generally copying figure (ex, hooks to/from pyplots global state and the GUI resources that a `canvas` may hold).

**Comment by derekbrokeit** [comment]
The reason a copy of a figure was considered is to make a temporary container for the axes in that figure. This figure will then be reconfigured to reduce margins and remove all text. The figure is only temporary and should not modify or mangle the original figure in trying to reduce spacing and text. If there is a better container or operation for this type of process, I would be very interested in hearing it. I am not familiar with the deep inner workings of `matplotlib`.

**Comment by QuLogic** [comment]
This was fixed in 3.2.0 by #15072.
